### PR TITLE
Skill exception handling template

### PR DESCRIPTION
**Pull request by lzc850612**
## Description
We're not handling exceptions for skills and it's causing problem such as multiple exceptions and no trace and logs. This change is to standardize the exception handling for skills.

## Related Issue
#279 

## Testing Steps
With contacts missing, from the auth scope string for AADv2, send the message 'send an email to Ryan letting him know I'll be back in Seattle on January 2nd' to the bot to trigger the exception about not having permission to find contact. With this change, there's only going to be one error message about error with the bot and developer being notified.

## Checklist
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by DingmaomaoBJTU** [comment]
Thank you so much for your kindly help!

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/SendEmail/SendEmailDialog.cs [hunk 1]
```diff
@@ -126,13 +131,34 @@ public async Task<DialogTurnResult> SendEmail(WaterfallStepContext sc, Cancellat
...
+            var resultString = result?.ToString();
+            if (!string.IsNullOrWhiteSpace(resultString) && resultString.Equals(CommonUtil.DialogTurnResultCancelAllDialogs, StringComparison.InvariantCultureIgnoreCase))
+            {
+                return outerDc.CancelAllDialogsAsync();
+            }
+            else
+            {
+                return base.EndComponentAsync(outerDc, result, cancellationToken);
+            }
+        }
```
Comment (DingmaomaoBJTU) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/SendEmail/SendEmailDialog.cs:162]: Can this move to some base class like ComponentDialog? Seems these functions are all the same, and maybe a base class implementation can help all other dialogs?

**Review by DingmaomaoBJTU — APPROVED** [review]
:shipit:

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/ConfirmRecipient/ConfirmRecipientDialog.cs [hunk 2]
```diff
@@ -327,8 +340,22 @@ public async Task<DialogTurnResult> AfterConfirmRecipient(WaterfallStepContext s
...
-                throw await HandleDialogExceptions(sc, ex);
+                await HandleDialogExceptions(sc, ex);
+
+                return new DialogTurnResult(DialogTurnStatus.Cancelled, CommonUtil.DialogTurnResultCancelAllDialogs);
+            }
+        }
+
+        protected override Task<DialogTurnResult> EndComponentAsync(DialogContext outerDc, object result, CancellationToken cancellationToken)
+        {
+            if (result.ToString().Equals(CommonUtil.DialogTurnResultCancelAllDialogs, StringComparison.InvariantCultureIgnoreCase))
```
Comment (feich-ms) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/ConfirmRecipient/ConfirmRecipientDialog.cs:351]: Seems missing "var resultString = result?.ToString(); if (!string.IsNullOrWhiteSpace(resultString) "?

**Review by feich-ms — APPROVED** [review]

**Review by darrenj — APPROVED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/ConfirmRecipient/ConfirmRecipientDialog.cs (hunk 2 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/ConfirmRecipient/ConfirmRecipientDialog.cs:351]: yes. i'll fix it in another PR

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/SendEmail/SendEmailDialog.cs (hunk 1 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/SendEmail/SendEmailDialog.cs:162]: yeah i think we should be able to move it to the lowest level of dialog.

**Pull request merged**